compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -241,14 +200,118 @@ safetyStep = \case
   ETypeRepF _ -> Safe 0
 
 
-infoStep :: ExprF Info -> Info
-infoStep e = Info (freeVarsStep (fmap freeVars e)) (safetyStep (fmap safety e))
+isTypeClassDictionary :: DefValue -> Bool
+isTypeClassDictionary DefValue{..}
+    = T.isPrefixOf "$f" (unExprValName (fst dvalBinder)) -- generic dictionary
+    || T.isPrefixOf "$d" (unExprValName (fst dvalBinder)) -- specialized dictionary
 
-simplifyExpr :: Expr -> Expr
-simplifyExpr = fst . cata go
+isTypeClassProjection :: DefValue -> Bool
+isTypeClassProjection DefValue{..} = go dvalBody
   where
+    go :: Expr -> Bool
+    go (ETyLam _ e) = go e
+    go (ETmLam _ e) = go e
+    go (ETmApp e _) = go e
+    go (EStructProj _ _) = True
+    go _ = False
+
+typeclassStep :: World -> ExprF TypeClassInfo -> TypeClassInfo
+typeclassStep world = \case
+    EValF x ->
+        case lookupValue x world of
+            Left _ -> TCNeither
+            Right dv
+                | isTypeClassProjection dv -> TCProjection (dvalBody dv)
+                | isTypeClassDictionary dv -> TCDictionary (dvalBody dv)
+                | otherwise -> TCNeither
+
+    ETyAppF tci ty ->
+        case tci of
+            TCProjection (ETyLam (x,_) e) ->
+                TCProjection (substExpr (typeSubst x ty) e)
+            TCDictionary (ETyLam (x,_) e) ->
+                TCDictionary (substExpr (typeSubst x ty) e)
+            _ ->
+                TCNeither
```

Comment:
Shouldn't we allow `ETmAppF` for dictionaries as well in order to inline dictionaries like `instance Eq a => Eq [a]`?

Reply:
Do these count as "known typeclasses"? AFAIK if you have a fixed `a`, GHC will specialize the typeclass, so we end up inlining them. But if you have a generic `a` then you aren't going to get inlining.

Reply:
I think we can perhaps leave this (& further experiments with inlining) to a separate PR.